src/validationinterface.cpp

```diff
@@ -406,49 +442,74 @@ void ThreadNotifyWallets(CBlockIndex *pindexLastTip)
+            // - we have more blocks that we _could_ add to batchScanners, and
+            //   are back below the memory limit (so should create more batches
+            //   before continuing, for optimal pipelining).
+            while (!(
+                blockStack.empty() ||
+                blockStack.rbegin() == blockStackScanned ||
+                (blockStackScanned != blockStack.rend() && belowBatchMemoryLimit())
+            )) {
+                auto& blockData = blockStack.back();
 
-            // Notify UI to display prev block's coinbase if it was ours.
-            static uint256 hashPrevBestCoinBase;
-            GetMainSignals().UpdatedTransaction(hashPrevBestCoinBase);
-            hashPrevBestCoinBase = block.vtx[0].GetHash();
+                // Read block from disk.
+                CBlock block;
+                if (!ReadBlockFromDisk(block, blockData.pindex, chainParams.GetConsensus())) {
+                    LogPrintf(
+                            "*** %s: Failed to read block %s while notifying wallets of block connects",
+                            __func__, blockData.pindex->GetBlockHash().GetHex());
+                    uiInterface.ThreadSafeMessageBox(
+                        strprintf(_("Error: A fatal internal error occurred, see %s for details"), GetDebugLogPath()),
+                        "", CClientUIInterface::MSG_ERROR);
+                    StartShutdown();
+                    return;
+                }
 
-            // This block is done!
-            pindexLastTip = blockData.pindex;
+                // Tell wallet about transactions that went from mempool
+                // to conflicted:
+                for (const CTransaction &tx : blockData.txConflicted) {
+                    SyncWithWallets(batchScanners, tx, NULL, blockData.pindex->nHeight + 1);
+                }
+                // ... and about transactions that got confirmed:
+                for (const CTransaction &tx : block.vtx) {
+                    SyncWithWallets(batchScanners, tx, &block, blockData.pindex->nHeight);
+                }
+                // Update cached incremental witnesses
+                // This will take the cs_main lock in order to obtain the CBlockLocator
+                // used by `SetBestChain`, but as that write only occurs once every
+                // WRITE_WITNESS_INTERVAL * 1000000 microseconds this should not be
+                // exploitable as a timing channel.
+                GetMainSignals().ChainTip(blockData.pindex, &block, blockData.oldTrees);
+
+                // Notify UI to display prev block's coinbase if it was ours.
+                static uint256 hashPrevBestCoinBase;
+                GetMainSignals().UpdatedTransaction(hashPrevBestCoinBase);
+                hashPrevBestCoinBase = block.vtx[0].GetHash();
+
+                // This block is done!
+                pindexLastTip = blockData.pindex;
+                blockStack.pop_back();
```

Comment:
```suggestion
                assert(blockStack.rbegin() != blockStackScanned);
                blockStack.pop_back();
```

Reply:
Isn't this redundant with the `while` condition? If we've entered the while loop then this holds by De Morgan.

Reply:
Yes it is redundant, but the code that allows you to infer that is 43 lines back.